pkg/platform/runtime/setup/buildlog/buildlog.go

```diff
@@ -67,69 +65,93 @@ func New(artifactMap map[artifact.ArtifactID]artifact.ArtifactRecipe, conn Build
 
 		var artifactErr error
 		for {
-			var msg message
+			var msg Message
 			err := conn.ReadJSON(&msg)
 			if err != nil {
 				errCh <- err
 				return
 			}
-			logging.Debug("Received response: " + msg.Type)
-
-			var artf artifact.ArtifactRecipe
-			var artifactMapped bool
-			if msg.ArtifactID != nil {
-				artf, artifactMapped = artifactMap[*msg.ArtifactID]
-			}
-			var artifactName string
-			if artifactMapped {
-				artifactName = artf.NameWithVersion()
-			}
+			logging.Debug("Received response: %d", msg.MessageType())
 
-			switch msg.Type {
-			case "build_failed":
-				errCh <- locale.WrapError(artifactErr, "err_logstream_build_failed", "Build failed with error message: {{.V0}}.", msg.Err())
+			switch msg.MessageType() {
+			case BuildFailed:
+				m := msg.messager.(BuildFailedMessage)
+				errCh <- locale.WrapError(artifactErr, "err_logstream_build_failed", "Build failed with error message: {{.V0}}.", m.ErrorMessage)
 				return
-			case "build_succeeded":
+			case BuildSucceeded:
 				return
-			case "artifact_started":
-				if !artifactMapped {
-					logging.Debug("build log streamer started unmapped artifact ID %s", msg.ArtifactID)
+			case ArtifactStarted:
+				m := msg.messager.(ArtifactMessage)
+				// NOTE: fix to ignore current noop "final pkg artifact"
+				if artifact.ArtifactID(m.ArtifactID) == recipeID {
 					continue
 				}
+				// ignore already built artifacts (they have been counted already)
+				if _, ok := alreadyBuilt[m.ArtifactID]; ok {
+					continue
+				}
+				events.ArtifactBuildStarting(m.ArtifactID)
+
+				// if verbose build logging is requested: Also subscribe to log messages for this artifacts
+				if os.Getenv(constants.LogBuildVerboseEnvVarName) == "true" {
```

Comment:
Should just store the result of this somewhere, it's not going to change. No need to do a `os.Getenv()` call every time.